draco/synthesis/mockcatalog.py

```diff
@@ -0,0 +1,1045 @@
+        # in each healpix pixel. These are uniform random numbers in [-0.5, 0.5],
+        # which will determine the source's relative displacement from the bin's
+        # central RA and dec.
+        if not self.srcs_at_pixel_centers:
+            rtheta = [np.random.uniform(size=num) - 0.5 for num in source_numbers]
+            rphi = [np.random.uniform(size=num) - 0.5 for num in source_numbers]
+
+        # Compute the square root of the angular pixel area,
+        # as a gross approximation of the pixel size.
+        ang_size = np.rad2deg(hp.nside2resol(self.nside))
+
+        # Redshifts corresponding to frequencies at bin centers
+        z_global = _freq_to_z(self.pdf.index_map["freq"][:])
+
+        # Number of sources in each rank
+        nsource_rank = np.sum([len(idxs[ii]) for ii in range(len(idxs))])
+
+        # Arrays to hold information on sources in local frequency section
+        mock_zs = np.empty(nsource_rank, dtype=np.float64)
+        mock_zerrs = np.empty(nsource_rank, dtype=np.float64)
+        mock_ra = np.empty(nsource_rank, dtype=np.float64)
+        mock_dec = np.empty(nsource_rank, dtype=np.float64)
+
+        # Loop over sources on this rank
+        source_count = 0
+        for zi in range(len(idxs)):  # For each local redshift bin
+            for si in range(len(idxs[zi])):  # For each source in z-bin zi
+
+                # Get dec, RA of center of pixel containing source
+                decbase, RAbase = _pix_to_radec(idxs[zi][si], self.nside)
+                # Get global index of z bin containing source, and central z
+                global_z_index = zi + self.lo
+                z_value = z_global["centre"][global_z_index]
+
+                # If desired, add random offset within z bin
+                if not self.z_at_channel_centers:
+                    z_value += z_global["width"][global_z_index] * rz[zi][si]
+
+                # If desired, add Gaussian z error
+                if self.sigma_z is not None:
+                    err = rzerr[zi][si] * self.sigma_z
+                    z_value += err
+                    mock_zerrs[source_count] = err
+                elif self.sigma_z_over_1plusz is not None:
+                    err = rzerr[zi][si] * self.sigma_z_over_1plusz * (1 + z_value)
+                    z_value += err
+                    mock_zerrs[source_count] = err
+                else:
+                    mock_zerrs[source_count] = 0
+
+                # Populate local arrays of source redshift, RA, dec,
+                # adding random angular offsets from pixel centers if desired
+                mock_zs[source_count] = z_value
+                mock_ra[source_count] = RAbase
+                mock_dec[source_count] = decbase
+                if not self.srcs_at_pixel_centers:
+                    mock_ra[source_count] += ang_size * rtheta[zi][si]
+                    mock_dec[source_count] += ang_size * rphi[zi][si]
+
+                source_count += 1
```

Comment:
I think this can all be done faster and more cleanly without the explicit loop of over sources within each redshift.

Reply:
i.e use vectorized routines and write the results to the right offsets in the `mock_(ra|dec|zs)` array.